File: mf/python/example_autoranging_impedance.py
# Copyright 2017 Zurich Instruments AG

"""
Zurich Instruments LabOne Python API Example

Demonstrate how to perform a manually triggered autoranging for impedance while working in
manual range mode. Sets the device to impedance manual range mode and execute
a single auto ranging event.

Requirements:
    * LabOne Version >= 20.02
    * Instruments:
        1 x Instrument with IA option
    * Connect Impedance Testfixture and attach a 1kOhm resistor

Usage:
    example_autoranging_impedance.py <device_id>
    example_autoranging_impedance.py -h | --help

Arguments:
    <device_id>  The ID of the device [device_type: MF*(IA)]

Options:
    -h --help                 Show this screen.

Raises:
    Exception     If the specified devices do not match the requirements.
    RuntimeError  If the devices is not "discoverable" from the API.

See the LabOne Programming Manual for further help:
https://docs.zhinst.com/labone_programming_manual/
"""
import time
import zhinst.utils


def run_example(device_id: str):
    """run the example."""

    apilevel_example = 6  # The API level supported by this example.
    # Call a zhinst utility function that returns:
    # - an API session `daq` in order to communicate with devices via the data server.
    # - the device ID string that specifies the device branch in the server's node hierarchy.
    # - the device's discovery properties.
    err_msg = "This example only supports instruments with IA option."
    (daq, device, _) = zhinst.utils.create_api_session(
        device_id, apilevel_example, required_options=["IA"], required_err_msg=err_msg
    )
    zhinst.utils.api_server_version_check(daq)

    # Create a base configuration: disable all available outputs, awgs, demods, scopes,...
    zhinst.utils.disable_everything(daq, device)

    # Now configure the instrument for this experiment. The following channels
    # and indices work on all devices with IA option. The values below may be
    # changed if the instrument has multiple IA modules.
    imp_index = 0
    curr_index = daq.getInt("/%s/imps/%d/current/inputselect" % (device, imp_index))
    volt_index = daq.getInt("/%s/imps/%d/voltage/inputselect" % (device, imp_index))
    man_curr_range = 10e-3
    man_volt_range = 10e-3
    exp_settings = [
        ["/%s/imps/%d/enable" % (device, imp_index), 1],
        ["/%s/imps/%d/mode" % (device, imp_index), 0],
        ["/%s/imps/%d/auto/output" % (device, imp_index), 1],
        ["/%s/imps/%d/auto/bw" % (device, imp_index), 1],
        ["/%s/imps/%d/freq" % (device, imp_index), 500],
        ["/%s/imps/%d/auto/inputrange" % (device, imp_index), 0],
        ["/%s/currins/%d/range" % (device, curr_index), man_curr_range],
        ["/%s/sigins/%d/range" % (device, volt_index), man_volt_range],
    ]
    daq.set(exp_settings)
    # Perform a global synchronisation between the device and the data server:
    # Ensure that the settings have taken effect on the device before setting
    # the next configuration.
    daq.sync()

    # After setting the device in manual ranging mode we want to trigger manually
    # a one time auto ranging to find a suitable range. Therefore, we trigger the
    # auto ranging for the current input as well as for the voltage input.
    trigger_auto_ranging = [
        ["/%s/currins/%d/autorange" % (device, curr_index), 1],
        ["/%s/sigins/%d/autorange" % (device, volt_index), 1],
    ]
    print("Start auto ranging. This takes a few seconds.")
    daq.set(trigger_auto_ranging)

    t_start = time.time()
    timeout = 20
    finished = 0
    # The auto ranging takes some time. We do not want to continue before the
    # best range is found. Therefore, we implement a loop to check if the auto
    # ranging is finished. These nodes maintain value 1 until autoranging has
    # finished.
    while not finished:
        time.sleep(0.5)
        currins_autorange = daq.getInt(
            "/%s/currins/%d/autorange" % (device, curr_index)
        )
        sigins_autorange = daq.getInt("/%s/sigins/%d/autorange" % (device, volt_index))
        # We are finished when both nodes have been set back to 0 by the device.
        finished = (currins_autorange == 0) and (sigins_autorange == 0)
        if time.time() - t_start > timeout:
            raise Exception(f"Autoranging failed after {timeout} seconds.")
    print(f"Auto ranging finished after {time.time() - t_start:0.1f} s.")

    auto_curr_range = daq.getDouble("/%s/currins/%d/range" % (device, curr_index))
    auto_volt_range = daq.getDouble("/%s/sigins/%d/range" % (device, volt_index))
    print(
        f"Current range changed from {man_curr_range:0.1e} A to {auto_curr_range:0.1e} A."
    )
    print(
        f"Voltage range changed from {man_volt_range:0.1e} V to {auto_volt_range:0.1e} V."
    )


if __name__ == "__main__":
    import sys
    from pathlib import Path

    cli_util_path = Path(__file__).resolve().parent / "../../utils/python"
    sys.path.insert(0, str(cli_util_path))
    cli_utils = __import__("cli_utils")
    cli_utils.run_commandline(run_example, __doc__)
    sys.path.remove(str(cli_util_path))

File: mf/matlab/example_autoranging_impedance.m
function example_autoranging_impedance(device_id, varargin)
% EXAMPLE_AUTORANGING_IMPEDANCE Demonstrate how to perform a manually triggered
% autoranging for impedance while working in manual range mode.
%
% USAGE EXAMPLE_AUTORANGING_IMPEDANCE(DEVICE_ID)
%
% The example sets the device to impedance manual range mode and executes a
% single auto ranging event.
% DEVICE_ID should be a string, e.g., 'dev1000'.
%
% NOTE Additional configuration: Connect Impedance Testfixture and attach a 1kOhm
% resistor
%
% NOTE Please ensure that the ziDAQ folders 'Driver' and 'Utils' are in your
% Matlab path. To do this (temporarily) for one Matlab session please navigate
% to the ziDAQ base folder containing the 'Driver', 'Examples' and 'Utils'
% subfolders and run the Matlab function ziAddPath().
% >>> ziAddPath;
%
% Use either of the commands:
% >>> help ziDAQ
% >>> doc ziDAQ
% in the Matlab command window to obtain help on all available ziDAQ commands.
%
% Copyright 2008-2016 Zurich Instruments AG

clear ziDAQ;

if ~exist('device_id', 'var')
    error(['No value for device_id specified. The first argument to the ' ...
           'example should be the device ID on which to run the example, ' ...
           'e.g. ''dev1000''.'])
end

% Check the ziDAQ MEX (DLL) and Utility functions can be found in Matlab's path.
if ~(exist('ziDAQ') == 3) && ~(exist('ziCreateAPISession', 'file') == 2)
    fprintf('Failed to either find the ziDAQ mex file or ziDevices() utility.\n')
    fprintf('Please configure your path using the ziDAQ function ziAddPath().\n')
    fprintf('This can be found in the API subfolder of your LabOne installation.\n');
    fprintf('On Windows this is typically:\n');
    fprintf('C:\\Program Files\\Zurich Instruments\\LabOne\\API\\MATLAB2012\\\n');
    return
end

% The API level supported by this example.
apilevel_example = 6;
% Create an API session; connect to the correct Data Server for the device.
[device, props] = ziCreateAPISession(device_id, apilevel_example);
ziApiServerVersionCheck();

% This example requires an MF device with IA option.
if ~ismember('IA', props.options)
    installed_options_str = ''; % Note: strjoin() unavailable in Matlab 2009.
    for option=props.options
      installed_options_str = [installed_options_str, ' ', option{1}];
    end
    error(['Required option set not satisfied. This example requires an MF ' ...
           'Instrument with the IA Option installed. Device `%s` reports ' ...
           'devtype `%s` and options `%s`.'], ...
          device, props.devicetype, installed_options_str);
end

% Create a base configuration: Disable all available outputs, awgs,
% demods, scopes,...
ziDisableEverything(device);

imp_c = '0';
curr_c = mat2str(ziDAQ('getInt', ['/' device '/imps/' imp_c  '/current/inputselect']));
volt_c = mat2str(ziDAQ('getInt', ['/' device '/imps/' imp_c  '/voltage/inputselect']));
man_curr_range = 10e-3;
man_volt_range = 10e-3;

%% Configure the device ready for this experiment.
% The following channels and indices work on all devices with IA option. The
% values below may be changed if the instrument has multiple IA modules.
ziDAQ('setInt', ['/' device '/imps/' imp_c '/enable'], 1);
ziDAQ('setInt', ['/' device '/imps/' imp_c '/mode'], 0);
ziDAQ('setInt', ['/' device '/imps/' imp_c '/auto/output'], 1);
ziDAQ('setInt', ['/' device '/imps/' imp_c '/auto/bw'], 1);
ziDAQ('setDouble', ['/' device '/imps/' imp_c '/freq'], 500);
ziDAQ('setInt', ['/' device '/imps/' imp_c '/auto/inputrange'], 0);
ziDAQ('setDouble', ['/' device '/currins/' curr_c '/range'], man_curr_range);
ziDAQ('setDouble', ['/' device '/sigins/' volt_c '/range'], man_volt_range);
ziDAQ('sync');

%% Trigger an autoranging
% Afater setting the device in impedance manual mode we trigger a single auto
% ranging event and wait until it is finished
fprintf('Start auto ranging. This might take a few seconds.\n');
ziDAQ('setInt', ['/' device '/currins/' curr_c '/autorange'], 1);
ziDAQ('setInt', ['/' device '/sigins/' volt_c '/autorange'], 1);
ziDAQ('sync');
currins_autorange = 1;
sigins_autorange = 1;
timeout = 20;
t0 = tic;
while currins_autorange || sigins_autorange
   if toc(t0) > timeout
       error('Autoranging failed to completed after %.2f s.\n', timeout);
   end
   pause(0.2);
   currins_autorange = ziDAQ('getInt', ['/' device '/currins/' curr_c '/autorange']);
   sigins_autorange = ziDAQ('getInt', ['/' device '/sigins/' volt_c '/autorange']);
end

fprintf('Auto ranging finished after %.1f s.\n', toc(t0));
auto_curr_range = ziDAQ('getDouble', ['/' device '/currins/' curr_c '/range']);
auto_volt_range = ziDAQ('getDouble', ['/' device '/sigins/' volt_c '/range']);
fprintf('Current range changed from %0.1eA to %0.1eA.\n', man_curr_range, auto_curr_range);
fprintf('Voltage range changed from %0.1eV to %0.1eV.\n', man_volt_range, auto_volt_range);
